### PR TITLE
get travis to build with both ghc 7.8 and 7.10

### PR DESCRIPTION
**Pull request by phunehehe**
https://github.com/commercialhaskell/stack/wiki/FAQ#how-do-i-use-this-with-travis-ci

### CODE REVIEW

**Comment by phunehehe** [comment]
okay stack didn't like that

**Comment by phunehehe** [comment]
It's kinda wasteful to build both on push and on PR

**Comment by proger** [comment]
Why do you use cabal for the other ghc?

**Comment by phunehehe** [comment]
Seems to work \o/

> Why do you use cabal for the other ghc?

because nix :)

**Comment by proger** [comment]
Does stack not work?

**Comment by phunehehe** [comment]
> Does stack not work?

Stack doesn't play really well with Nix. Also I feel like I should be able to build my tools with whatever I want.

Do you want/plan to stop supporting cabal? If so, deleting the cabal file would make sense, and I'll figure out something else.

**Comment by proger** [comment]
Stack is using those files, so we can't drop them at this point. It just feels weird to use two different frontend tools to run builds for different compilers.

**Comment by phunehehe** [comment]
If we still need cabal, then let's do cabal right. Having cabal checks helps us do cabal right :)
Basically, with this PR, we make sure that we can build using either cabal or stack. We could also try harder and make sure that we can build using (cabal, slack) * (GHC 7.4, GHC 7.10), but I assume we don't need that just yet.

**Comment by proger** [comment]
The problem with using cabal-install is that it assumes that constraints in *.cabal files are set properly, whereas in our case they are not (I rely on stack curated package sets to do this work for me). So the fact that cabal builds work out just fine is a surprise for me. :)

**Comment by sol** [comment]
No upper bounds is a sane default (contrary to popular...

**Comment by phunehehe** [comment]
I don't feel very good about the upper bound for `base` either. I could drop it and remove the `cabal check`. Would that be enough to get this merged?

**Comment by sol** [comment]
The upper bounds on base I've would leave (for the sake of making cabal happy)

**Comment by phunehehe** [comment]
Then all is good, it seems? I'll also try converting the thing to hpack.

File: .travis.yml [hunk 1]
```diff
@@ -1,19 +1,40 @@
...
-- cp ~/.local/bin/upcast .
-- strip -p --strip-unneeded --remove-section=.comment upcast
-- upx upcast
+- |
+   set -ex
+   if [ -n "$USE_STACK" ]
+   then
+     $STACK build --ghc-options "$GHC_OPTIONS"
+     $STACK install --ghc-options "$GHC_OPTIONS"
+     cp ~/.local/bin/upcast .
```
Comment (proger) [inline .travis.yml]: i think you need to make these steps run regardless of stack/cabal

Comment (phunehehe) [inline .travis.yml]: But, Stack doesn't like a GHC version different than the one it knows about, so it doesn't really make sense to build GHC 7.10 with Stack. Or does it?

**Comment by phunehehe** [comment]
https://github.com/zalora/microgram/pull/23

File: .travis.yml (hunk 1 above)
Comment (proger) [inline .travis.yml]: > GHC 7.10 with Stack

ghc-7.10 will work out of the box if you use a different resolver (you'll need to specify a different stack file, see [amazonka for example](https://github.com/brendanhay/amazonka/blob/develop/script/travis-lifecycle-install#L14))

however, my point here was to strip/upx the upcast binary regardless of whatever you build it with, so the artifact uploader handles it

Comment (phunehehe) [inline .travis.yml]: adapted, should be good to go

**Comment by proger** [comment]
Btw why do we need to support ghc 7.8?..

**Comment by proger** [comment]
Old nixpkgs i suppose?

**Comment by phunehehe** [comment]
Nope, I don't care about 7.8 at all. I thought you might want to stick to a stable stackage snapshot. If that's not the case 7.8 can be removed.

**Pull request merged**